dali/c_api_2/data_objects.h

```diff
@@ -0,0 +1,657 @@
+    tl_->SetSize(num_samples);
+    tl_->set_type(dtype);
+    tl_->set_sample_dim(ndim);
+    tl_->SetLayout(new_layout);
+
+    auto deletion_order = tl_->order();
+
+    auto type_info = TypeTable::GetTypeInfo(dtype);
+    auto element_size = type_info.size();
+    for (int i = 0; i < num_samples; i++) {
+      TensorShape<> sample_shape(make_cspan(samples[i].shape, samples[i].ndim));
+      size_t sample_bytes = volume(sample_shape) * element_size;
+      std::shared_ptr<void> sample_ptr;
+      if (sample_deleters) {
+        sample_ptr = std::shared_ptr<void>(
+          samples[i].data,
+          BufferDeleter{sample_deleters[i], deletion_order});
+      } else {
+        sample_ptr = std::shared_ptr<void>(samples[i].data, [](void*) {});
+      }
+
+      tl_->SetSample(
+        i,
+        sample_ptr,
+        sample_bytes,
+        tl_->is_pinned(),
+        sample_shape,
+        dtype,
+        tl_->device_id(),
+        tl_->order(),
+        new_layout);
+    }
+  }
+
+  daliBufferPlacement_t GetBufferPlacement() const override {
+    daliBufferPlacement_t placement;
+    placement.device_id = tl_->device_id();
+    StorageDevice dev = backend_to_storage_device<Backend>::value;
+    placement.device_type = static_cast<daliStorageDevice_t>(dev);
+    placement.pinned = tl_->is_pinned();
+    return placement;
+  }
+
+  void SetStream(std::optional<cudaStream_t> stream, bool synchronize) override {
+    tl_->set_order(stream.has_value() ? AccessOrder(*stream) : AccessOrder::host(), synchronize);
+  }
+
+  void SetLayout(const char *layout_string) override {
+    if (layout_string) {
+      TensorLayout layout(layout_string);
+      Validate(layout, tl_->sample_dim());
+      tl_->SetLayout(layout);
+    } else {
+      tl_->SetLayout("");
+    }
+  }
+
+  const char *GetLayout() const override {
+    auto &layout = tl_->GetLayout();
+    return !layout.empty() ? layout.data() : nullptr;
+  }
+
+  std::optional<cudaStream_t> GetStream() const override {
+    auto o = tl_->order();
+    if (o.is_device())
+      return o.stream();
+    else
+      return std::nullopt;
+  }
+
+  std::optional<cudaEvent_t> GetReadyEvent() const override {
+    auto &e = tl_->ready_event();
+    if (e)
+      return e.get();
+    else
+      return std::nullopt;
+  }
+
+  cudaEvent_t GetOrCreateReadyEvent() override {
+    auto &e = tl_->ready_event();
+    if (e)
+      return e.get();
+    int device_id = tl_->device_id();
+    if (device_id < 0)
+      throw std::runtime_error("The tensor list is not associated with a CUDA device.");
+    tl_->set_ready_event(CUDASharedEvent::Create(device_id));
+    return tl_->ready_event().get();
+  }
+
+  daliTensorDesc_t GetTensorDesc(int sample) const override {
+    auto &shape = tl_->shape();
+    if (sample < 0 || sample >= shape.num_samples())
+      throw std::out_of_range(make_string("The sample index ", sample, " is out of range. "
+        "Valid indices are [0..", shape.num_samples() - 1, "]."));
```

Comment:
Nitpick: This will look strange for 0 samples

Reply:
Fixed.